tests/frontier/scenarios/common.py

```diff
@@ -0,0 +1,302 @@
+    gaslimit: int
+
+
+@dataclass
+class ScenarioGeneratorInput:
+    """
+    Parameters for the scenario generator function.
+
+    Attributes:
+        fork (Fork): Fork for which we ask to generate scenarios
+        pre (Alloc): Access to the state to be able to deploy contracts into pre
+        operation (Bytecode): Evm bytecode program that will be tested
+        external_address (Address): Static external address for ext opcodes
+
+    """
+
+    fork: Fork
+    pre: Alloc
+    operation_code: Bytecode
+    external_address: Address
+    external_balance: int
+
+
+@dataclass
+class Scenario:
+    """
+    Describe test scenario that will be run in test for each program.
+
+    Attributes:
+        name (str): Scenario name for the test vector
+        code (Address): Address that is an entry point for scenario code
+        env (ScenarioEnvironment): Evm values for ScenarioExpectAddress map
+        reverting (bool): If scenario reverts program execution, making result 0 (default: False)
+
+    """
+
+    name: str
+    code: Address
+    env: ScenarioEnvironment
+    halts: bool = False
+
+
+def translate_result(
+    res: ProgramResult, env: ScenarioEnvironment, exec_env: ExecutionEnvironment
+) -> int:
+    """
+    Translate expected program result code into concrete value,
+    given the scenario evm environment and test execution environment.
+    """
+    if exec_env.fork < res.from_fork:
+        return 0
+    if not res.static_support and env.has_static:
+        return 0
+    if isinstance(res.result, ScenarioExpectOpcode):
+        if res.result == ScenarioExpectOpcode.TX_ORIGIN:
+            return int(exec_env.origin.hex(), 16)
+        if res.result == ScenarioExpectOpcode.CODE_ADDRESS:
+            return int(env.code_address.hex(), 16)
+        if res.result == ScenarioExpectOpcode.CODE_CALLER:
+            return int(env.code_caller.hex(), 16)
+        if res.result == ScenarioExpectOpcode.BALANCE:
+            return int(env.ext_balance)
+        if res.result == ScenarioExpectOpcode.CALL_VALUE:
+            return int(env.call_value)
+        if res.result == ScenarioExpectOpcode.CALL_DATALOAD_0:
+            return env.call_dataload_0
+        if res.result == ScenarioExpectOpcode.CALL_DATASIZE:
+            return env.call_datasize
+        if res.result == ScenarioExpectOpcode.GASPRICE:
+            return exec_env.gasprice
+        if res.result == ScenarioExpectOpcode.COINBASE:
+            return int(exec_env.coinbase.hex(), 16)
+        if res.result == ScenarioExpectOpcode.TIMESTAMP:
+            return exec_env.timestamp
+        if res.result == ScenarioExpectOpcode.NUMBER:
+            return exec_env.number
+        if res.result == ScenarioExpectOpcode.GASLIMIT:
+            return exec_env.gaslimit
+        if res.result == ScenarioExpectOpcode.SELFBALANCE:
+            return int(env.selfbalance)
+    else:
+        return res.result
+    return 0
+
+
+def replace_special_calls_in_operation(
```

Comment:
I think we should not use patterns like this in my opinion, we are basically making code introspection/morphing code we produced ourselves.

We should generate the correct code from the start.

Perhaps make the `operation` a callable code generator or just use fixtures.

Reply:
This is because scenario program does not have a constructor, so it uses cistom system contacts